cudax/include/cuda/experimental/__utility/basic_any/rtti.cuh

```diff
@@ -0,0 +1,251 @@
+    // addresses and, if that fails, it does a string comparison. What we want is
+    // to check _all_ the addresses first, and only if they all fail, resort to
+    // string comparisons. So do two passes over the __base_vptr_map.
+    constexpr _CUDA_VSTD::__type_info_ref __id = _CCCL_TYPEID(_Interface);
+
+    for (size_t __i = 0; __i < __nbr_interfaces_; ++__i)
+    {
+      if (&__id == __base_vptr_map_[__i].__typeid_)
+      {
+        return static_cast<__vptr_for<_Interface>>(__base_vptr_map_[__i].__vptr_);
+      }
+    }
+
+    for (size_t __i = 0; __i < __nbr_interfaces_; ++__i)
+    {
+      if (__id == *__base_vptr_map_[__i].__typeid_)
+      {
+        return static_cast<__vptr_for<_Interface>>(__base_vptr_map_[__i].__vptr_);
+      }
+    }
+
+    return nullptr;
+  }
+
+  void (*__dtor_)(void*, bool) noexcept;
+  __object_metadata const* __object_info_;
+  _CUDA_VSTD::__type_info_ptr __interface_typeid_ = nullptr;
+  __base_info const* __base_vptr_map_;
+};
+
+template <size_t _NbrInterfaces>
+struct __rtti_ex : __rtti
+{
+  template <class _Tp, class _Super, class... _Interfaces, class _VPtr>
+  _CUDAX_HOST_API constexpr __rtti_ex(__tag<_Tp, _Super> __type, __tag<_Interfaces...> __ibases, _VPtr __self) noexcept
+      : __rtti{__type, __ibases, __base_vptr_array}
+      , __base_vptr_array{{&_CCCL_TYPEID(_Interfaces), static_cast<__vptr_for<_Interfaces>>(__self)}...}
+  {}
+
+  __base_info __base_vptr_array[_NbrInterfaces];
+};
+
+//!
+//! __try_vptr_cast
+//!
+//! This function ignores const qualification on the source and destination
+//! interfaces.
+//!
+template <class _SrcInterface, class _DstInterface>
+_CCCL_NODISCARD _CUDAX_HOST_API auto
+__try_vptr_cast(__vptr_for<_SrcInterface> __src_vptr) noexcept -> __vptr_for<_DstInterface>
+{
+  static_assert(_CUDA_VSTD::is_class_v<_SrcInterface> && _CUDA_VSTD::is_class_v<_DstInterface>, "expected class types");
+  if (__src_vptr == nullptr)
+  {
+    return nullptr;
+  }
+  else if constexpr (_CUDA_VSTD::is_same_v<_SrcInterface const, _DstInterface const>)
+  {
+    return __src_vptr;
+  }
+  else if constexpr (extension_of<_SrcInterface, _DstInterface>)
+  {
+    //! Fast up-casts:
+    return __src_vptr->__query_interface(_DstInterface());
+  }
+  else
+  {
+    //! Slow down-casts and cross-casts:
+    __rtti const* rtti = __src_vptr->__query_interface(iunknown());
+    return rtti->__query_interface(_DstInterface());
+  }
+}
+
+template <class _SrcInterface, class _DstInterface>
+_CCCL_NODISCARD _CUDAX_HOST_API auto __vptr_cast(__vptr_for<_SrcInterface> __src_vptr) //
+  noexcept(_CUDA_VSTD::is_same_v<_SrcInterface, _DstInterface>) //
+  -> __vptr_for<_DstInterface>
+{
+  if constexpr (_CUDA_VSTD::is_same_v<_SrcInterface, _DstInterface>)
+  {
+    return __src_vptr;
+  }
+  else
+  {
+    auto __dst_vptr = __try_vptr_cast<_SrcInterface, _DstInterface>(__src_vptr);
+    if (!__dst_vptr && __src_vptr)
+    {
+      __throw_bad_any_cast();
+    }
+    return __dst_vptr;
+  }
```

Comment:
This might need an `_CCCL_UNREACHABLE()` because of if constexpr handling in older compilers